File: Python/Tree/M-437. Path Sum III.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    @classmethod
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        path = 0

        def dfs(node: Optional[TreeNode], node_vals: list[int]):
            nonlocal path

            if node is None:
                return

            cur_node_vals = node_vals.copy()
            cur_node_vals.append(node.val)
            cur_path_sum = sum(cur_node_vals)
            if cur_path_sum == targetSum:
                path += 1
            if len(cur_node_vals) > 1:
                cur_path_sum_copy = cur_path_sum
                for i in range(0, len(cur_node_vals) - 1):
                    cur_path_sum_copy -= cur_node_vals[i]
                    if cur_path_sum_copy == targetSum:
                        path += 1
            dfs(node.left, cur_node_vals)
            dfs(node.right, cur_node_vals)

        dfs(root, [])
        return path

File: Tree/M-437. Path Sum III/M-437. Path Sum III.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number {
  let path = 0;
  const dfs = (node: TreeNode | null, nodeVals: number[]): void => {
    if (!node) return;

    const curNodeVals = [...nodeVals];
    curNodeVals.push(node.val);
    const curSum = curNodeVals.reduce((sum, val) => sum + val, 0);
    if (curSum === targetSum) path++;
    if (curNodeVals.length > 1) {
      let copyCurSum = curSum;
      /** start from 1 since root is not removeable */
      for (let i = 0; i < curNodeVals.length - 1; i++) {
        copyCurSum -= curNodeVals[i];
        if (copyCurSum === targetSum) path++;
      }
    }

    dfs(node.left, curNodeVals);
    dfs(node.right, curNodeVals);
  };

  dfs(root, []);
  return path;
}
